File: 2021년/8월/0809 쿼드트리.cpp
#include <iostream>
#include <string>
using namespace std;

string maps[64];

void solution(int m, int x, int y)
{
	char color = maps[x][y];
	for(int i=x;i<x+m;i++)
	{
		for(int j=y;j<y+m;j++)
		{
			if(color != maps[i][j])
			{
				cout << '(';
				solution(m/2, x, y);
				solution(m/2, x, y+m/2);
				solution(m/2, x+m/2, y);
				solution(m/2, x+m/2, y+m/2);
				cout << ')';
				return;
			}
		}
	}
	
	if(color == '0') cout << '0';
	else cout << '1';
}

int main()
{
	int n;
	cin >> n;
	for(int i=0;i<n;i++) cin >> maps[i];
	
	solution(n, 0, 0);
}

File: 2021년/8월/0809 쿼드트리.py
# 백준_1992_쿼드트리_분할 정복_실버 1

import sys
input = sys.stdin.readline


def solution(m, x, y):
    color = maps[x][y]
    for i in range(x, x+m):
        for j in range(y, y+m):
            if color != maps[i][j]:
                result.append('(')
                solution(m//2, x, y)
                solution(m//2, x, y+m//2)
                solution(m//2, x+m//2, y)
                solution(m//2, x+m//2, y+m//2)
                result.append(')')
                return

    if color == '0':
        result.append('0')
    else:
        result.append('1')


n = int(input())
maps = []
for _ in range(n):
    maps.append(list(input().strip()))

result = []
solution(n, 0, 0)
for i in range(len(result)):
    print(result[i], end='')
